File: src/main/java/cn/edu/sxu/museai/controller/AppController.java
package cn.edu.sxu.museai.controller;

import cn.edu.sxu.museai.aop.annotations.AuthCheck;
import cn.edu.sxu.museai.common.BaseResponse;
import cn.edu.sxu.museai.common.ResultUtils;
import cn.edu.sxu.museai.model.dto.*;
import cn.edu.sxu.museai.model.dto.app.*;
import cn.edu.sxu.museai.model.entity.User;
import cn.edu.sxu.museai.model.enums.UserRoleEnum;
import cn.edu.sxu.museai.model.vo.AppVO;
import cn.edu.sxu.museai.service.AppService;
import cn.edu.sxu.museai.service.UserService;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;

import java.util.List;

/**
 * 应用 控制层
 */
@RestController
@RequestMapping("/app")
@RequiredArgsConstructor
public class AppController {

    private final AppService appService;
    private final UserService userService;

    @PostMapping(value = "/chat",  produces = MediaType.TEXT_EVENT_STREAM_VALUE)
    public Flux<String> chatToGenApp(@RequestBody AppChatRequest appChatRequest, HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        return appService.chatToGenApp(appChatRequest.getUserMessage(), appChatRequest.getAppId(), loginUser.getId());
    }

    @PostMapping("/deploy")
    public BaseResponse<String> deployApp(@RequestBody AppDeployRequest appDeployRequest, HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        String path = appService.deploy(appDeployRequest.getAppId() ,loginUser.getId());

        return null;
    }
    /**
     * 普通用户根据提示词创建应用
     * @param appAddRequest 应用创建请求
     * @param request http请求
     * @return 应用id
     */
    @PostMapping("/create")
    @AuthCheck(UserRoleEnum.USER)
    public BaseResponse<Long> createApp(@RequestBody AppAddRequest appAddRequest, HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        Long appId = appService.createApp(appAddRequest, loginUser.getId());
        return ResultUtils.success(appId);
    }

    /**
     * 普通用户修改应用名称
     * @param appNameUpdateRequest 应用名称更新请求
     * @param request http请求
     * @return 是否成功
     */
    @AuthCheck(UserRoleEnum.USER)
    @PostMapping("/update/name")
    public BaseResponse<Boolean> updateAppName(@RequestBody AppNameUpdateRequest appNameUpdateRequest, HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        Boolean result = appService.updateAppName(appNameUpdateRequest, loginUser.getId());
        return ResultUtils.success(result);
    }

    /**
     * 普通用户删除自己的应用
     * @param id 应用id
     * @param request http请求
     * @return 是否成功
     */
    @AuthCheck(UserRoleEnum.USER)
    @PostMapping("/delete")
    public BaseResponse<Boolean> deleteApp(@RequestParam("id") Long id, HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        Boolean result = appService.deleteApp(id, loginUser.getId());
        return ResultUtils.success(result);
    }

    /**
     * 普通用户查看自己的应用详情
     * @param id 应用id
     * @param request http请求
     * @return 应用详情
     */
    @AuthCheck(UserRoleEnum.USER)
    @GetMapping("/get")
    public BaseResponse<AppVO> getAppDetail(@RequestParam("id") Long id, HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        AppVO appVO = appService.getAppDetail(id, loginUser.getId());
        return ResultUtils.success(appVO);
    }

    /**
     * 普通用户分页查询自己的应用列表
     * @param appQueryRequest 查询请求
     * @param request http请求
     * @return 应用列表
     */
    @AuthCheck(UserRoleEnum.USER)
    @GetMapping("/list/my")
    public BaseResponse<List<AppVO>> listMyApps(AppQueryRequest appQueryRequest, HttpServletRequest request) {
        User loginUser = userService.getLoginUser(request);
        List<AppVO> appList = appService.listUserApps(appQueryRequest, loginUser.getId());
        return ResultUtils.success(appList);
    }

    /**
     * 查询精选应用列表，按优先级排序
     * @param appQueryRequest 查询请求
     * @return 应用列表
     */
    @GetMapping("/list/featured")
    public BaseResponse<List<AppVO>> listFeaturedApps(AppQueryRequest appQueryRequest) {
        List<AppVO> appList = appService.listFeaturedApps(appQueryRequest);
        return ResultUtils.success(appList);
    }

    /**
     * 管理员删除任意应用
     * @param deleteRequest 应用id
     * @return 是否成功
     */
    @AuthCheck(UserRoleEnum.ADMIN)
    @PostMapping("/admin/delete")
    public BaseResponse<Boolean> deleteAppByAdmin(@RequestBody DeleteRequest deleteRequest) {
        Boolean result = appService.deleteAppByAdmin(deleteRequest);
        return ResultUtils.success(result);
    }

    /**
     * 管理员更新应用信息
     * @param appUpdateRequest 应用更新请求
     * @return 是否成功
     */
    @AuthCheck(UserRoleEnum.ADMIN)
    @PostMapping("/admin/update")
    public BaseResponse<Boolean> updateAppByAdmin(@RequestBody AppUpdateRequest appUpdateRequest) {
        Boolean result = appService.updateAppByAdmin(appUpdateRequest);
        return ResultUtils.success(result);
    }

    /**
     * 管理员分页查询应用列表
     * @param appQueryRequest 查询请求
     * @return 应用列表
     */
    @AuthCheck(UserRoleEnum.ADMIN)
    @GetMapping("/admin/list")
    public BaseResponse<List<AppVO>> listAppsByAdmin(AppQueryRequest appQueryRequest) {
        List<AppVO> appList = appService.listAppsByAdmin(appQueryRequest);
        return ResultUtils.success(appList);
    }

    /**
     * 管理员查看应用详情
     * @param id 应用id
     * @return 应用详情
     */
    @AuthCheck(UserRoleEnum.ADMIN)
    @GetMapping("/admin/get")
    public BaseResponse<AppVO> getAppDetailByAdmin(@RequestParam("id") Long id) {
        AppVO appVO = appService.getAppDetailByAdmin(id);
        return ResultUtils.success(appVO);
    }
}

File: muse-ai-front/src/api/appController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/request'

/** 此处后端没有提供注释 POST /app/admin/delete */
export async function deleteAppByAdmin(body: API.DeleteRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/app/admin/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /app/admin/get */
export async function getAppDetailByAdmin(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getAppDetailByAdminParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseAppVO>('/app/admin/get', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /app/admin/list */
export async function listAppsByAdmin(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listAppsByAdminParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePageResultAppVO>('/app/admin/list', {
    method: 'GET',
    params: {
      ...params,
      appQueryRequest: undefined,
      ...params['appQueryRequest'],
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /app/admin/pin */
export async function pinApp(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.pinAppParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/app/admin/pin', {
    method: 'POST',
    params: {
      ...params,
      appPinRequest: undefined,
      ...params['appPinRequest'],
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /app/admin/update */
export async function updateAppByAdmin(
  body: API.AppUpdateRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/app/admin/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /app/chat */
export async function chatToGenApp(body: API.AppChatRequest, options?: { [key: string]: any }) {
  return request<string[]>('/app/chat', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /app/create */
export async function createApp(body: API.AppAddRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseLong>('/app/create', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /app/delete */
export async function deleteApp(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.deleteAppParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/app/delete', {
    method: 'POST',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /app/deploy */
export async function deployApp(body: API.AppDeployRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseString>('/app/deploy', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /app/download */
export async function downloadApp(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.downloadAppParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseString>('/app/download', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /app/get */
export async function getAppDetail(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getAppDetailParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseAppVO>('/app/get', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /app/list/featured */
export async function listFeaturedApps(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listFeaturedAppsParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePageResultAppVO>('/app/list/featured', {
    method: 'GET',
    params: {
      ...params,
      appQueryRequest: undefined,
      ...params['appQueryRequest'],
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /app/list/my */
export async function listMyApps(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.listMyAppsParams,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePageResultAppVO>('/app/list/my', {
    method: 'GET',
    params: {
      ...params,
      appQueryRequest: undefined,
      ...params['appQueryRequest'],
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /app/update/name */
export async function updateAppName(
  body: API.AppNameUpdateRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/app/update/name', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}
